feat(files_sharing): reminder for link shares with expiration date

Luka-sama opened the pull request:
* Resolves: #45676

## Summary

This pull request implements the feature as described in the associated issue. A cron job runs hourly to check if any empty folders have less than a day remaining before their expiration date. The feature supports two types of shares: user shares and email shares.

Additionally, a migration was added to include a 'reminder sent' flag, ensuring that reminders are neither skipped nor sent multiple times when the job runs at irregular intervals, especially when using webcron or ajax instead of cron.

This solution can be easily extended to send reminders for link shares without an expiration date (e.g. after 4 days), if the maintainers choose to implement this feature.

![grafik](https://github.com/user-attachments/assets/e74c9483-3382-45ef-a1df-61cd50475a14)

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

Luka-sama committed:
feat(files_sharing): reminder for link shares with expiration date

Signed-off-by: Stefan Cherniakov <luka-sama@pm.me>

come-nc commented on apps/files_sharing/lib/SharesReminderJob.php:
```diff
@@ -0,0 +1,261 @@
+		$maxDate = new \DateTime();
+		$maxDate->setTimestamp($maxDate->getTimestamp() + self::SECONDS_BEFORE_REMINDER);
+
+		$qb = $this->db->getQueryBuilder();
+		$qb->select('id', 'share_type')
+			->from('share')
+			->where(
+				$qb->expr()->andX(
+					$qb->expr()->orX(
+						$qb->expr()->eq('share_type', $qb->expr()->literal(IShare::TYPE_USER)),
+						$qb->expr()->eq('share_type', $qb->expr()->literal(IShare::TYPE_EMAIL))
+					),
+					$qb->expr()->eq('item_type', $qb->expr()->literal('folder')),
+					$qb->expr()->gte('expiration', $qb->createNamedParameter($minDate->format('Y-m-d H:i:s'))),
+					$qb->expr()->lt('expiration', $qb->createNamedParameter($maxDate->format('Y-m-d H:i:s'))),
+					$qb->expr()->eq('reminder_sent', $qb->createNamedParameter(
+						false, IQueryBuilder::PARAM_BOOL
+					))
+				)
+			);
```
Should this go into a mapper or manager of some kind? I know in other apps we usually do not hit the DB directly like that, not sure for shares.

Luka-sama replied:
I could move this to the share manager, but I'm not sure if I should, since the query is very specific to this particular job 🤔

come-nc commented on apps/files_sharing/appinfo/info.xml:
```diff
@@ -14,7 +14,7 @@
 
 	</description>
-	<version>1.23.0</version>
+	<version>1.23.1</version>
```
That’s needed for the migration to run I suppose?

Luka-sama replied:
Yes, with a new version number the user will be required to do "occ upgrade" (via web interface or console). Also, it seems logical to me to update the version when the new feature comes out.
But of course I could be wrong as I'm new to nextcloud and don't know all the details of how it works here.

come-nc commented on apps/files_sharing/lib/SharesReminderJob.php:
```diff
@@ -0,0 +1,261 @@
+			$qb->update('share')
+				->where($qb->expr()->eq('id', $qb->createNamedParameter($share->getId())))
+				->set('reminder_sent', $qb->createNamedParameter($share->getReminderSent()))
+				->execute();
```
```suggestion
				->executeStatement();
```
But again, I’d like to see that moved somewhere else.
The updateShare method seems a bit much for such a small flag change though, so I don’t know :thinking:

Luka-sama replied:
I think updateShare works well here, the overhead (if any) is quite insignificant.

come-nc commented on lib/private/Share20/Share.php:
```diff
@@ -72,6 +72,8 @@ class Share implements IShare {
 	private $nodeCacheEntry;
 	/** @var bool */
 	private $hideDownload = false;
+	/** @var bool */
+	private $reminderSent = false;
```
```suggestion
	private bool $reminderSent = false;
```

Luka-sama replied:
Should I maybe update other properties to this format as well (to make it consistent)?

Luka-sama replied:
> Should I maybe update other properties to this format as well (to make it consistent)?

I decided against it because it breaks a lot of code (e.g. some int properties are sometimes set to null). So I did the strict typing only for my property.

come-nc commented on lib/private/Share20/Share.php:
```diff
@@ -613,4 +615,13 @@ public function setHideDownload(bool $hide): IShare {
+		return $this;
+	}
+
+	public function getReminderSent(): bool {
+		return $this->reminderSent;
+	}
```
But is the information fetch from DB and put into this object by the manager or the providers?

Luka-sama replied:
Now it does, thanks for noticing!

come-nc commented on apps/files_sharing/lib/SharesReminderJob.php:
```diff
@@ -0,0 +1,261 @@
+	 * @param array $argument unused argument
+	 * @throws Exception if a database error occurs
+	 */
+	public function run(mixed $argument): void {
+		$shares = $this->getShares();
+		[$foldersByEmail, $langByEmail] = $this->prepareReminders($shares);
+		$this->sendReminders($foldersByEmail, $langByEmail);
+	}
```
You need to put a limit here, and work only on a chunk of possible reminders, like `1000`. Otherwise you can run into memory issues and such.
Also, you are first gathering all the info and then sending all the email, consider using Generators to send shares as you find them.
Also, you need to consider what happens if your job crashes or is stopped mid-process. In the current state it sets the flag reminder_sent on all shares first, which means it won’t get sent for other shares if it crashes on a share before. This is better than re-sending reminders, but it would still be better to flag shares one by one just before sending the email.

Luka-sama replied:
Fixed, thank you!

skjnldsv commented:
Great idea @Luka-sama !!
Take your time to address Come's comments :+1:

Luka-sama commented:
Thank you @come-nc and @skjnldsv! The idea comes from the issue, but I found it a very interesting task to implement it here :)
That's a very nice idea with generators! I will fix/address this and other problems in the next days.

Luka-sama committed:
fix(files_sharing): Make share reminders more stable & fix issues

Signed-off-by: Stefan Cherniakov <luka-sama@pm.me>

Luka-sama commented:
So, I've fixed the problems that @come-nc found!
To do this, I moved the condition checking (write permissions and folder emptiness) to the database query - to avoid the situation where the job checks the same 1000 shares every time.

come-nc commented on apps/files_sharing/lib/SharesReminderJob.php:
```diff
@@ -0,0 +1,213 @@
+	 */
+	public function run(mixed $argument): void {
+		foreach ($this->getShares() as $share) {
+			$reminderInfo = $this->prepareReminder($share);
+			$this->sendReminder($reminderInfo);
```
```suggestion
			$reminderInfo = $this->prepareReminder($share);
			if ($reminderInfo !== null) {
				$this->sendReminder($reminderInfo);
			}
```

come-nc commented on apps/sharebymail/lib/ShareByMailProvider.php:
```diff
@@ -755,6 +755,7 @@ public function update(IShare $share, ?string $plainTextPassword = null): IShare
 			->set('hide_download', $qb->createNamedParameter((int)$share->getHideDownload(), IQueryBuilder::PARAM_INT))
 			->set('attributes', $qb->createNamedParameter($shareAttributes))
 			->set('mail_send', $qb->createNamedParameter((int)$share->getMailSend(), IQueryBuilder::PARAM_INT))
+			->set('reminder_sent', $qb->createNamedParameter((int)$share->getReminderSent(), IQueryBuilder::PARAM_INT))
```
Why not PARAM_BOOL?

Luka-sama replied:
I guess I was following the example of other fields that mostly use int for boolean data. However, some fields use bool – and since the tests passed, I've changed to bool as well.

come-nc reviewed (commented):
`reminder_sent` should be a bool in DB as well, otherwise good.
Only thing a bit questionable is how would other providers add support for this, but this can be refactored in later.

Luka-sama committed:
fix(files_sharing): Add missing check for null & use bool instead of int for reminder_sent field

Signed-off-by: Stefan Cherniakov <luka-sama@pm.me>

Luka-sama commented:
Thank you for your code review!
Yes, currently only two types of shares are supported (both in providers and in the job itself). However, support for other types can easily be added if needed.

come-nc reviewed (approved):
What I meant was that this could’ve been added in a more extensible way, for instance with an interface for providers which want to implement the feature, and which adds a method to get which email address the reminder should be sent to.
Just thinking out loud, I have no idea if this would’ve been a good idea and all it implies, just a suggestion for future work to think about extensibility by applications.

Approved :+1:

Luka-sama committed:
fix(files_sharing): Add check for null

Signed-off-by: Stefan Cherniakov <luka-sama@pm.me>

Luka-sama commented:
You're right, that's definitely a good idea for the future to make the feature more universal this way.

Thanks!

skjnldsv commented:
Ok, let's go! :rocket:

Pytal commented:
This broke CI on master cc @skjnldsv

susnux commented on apps/files_sharing/lib/Migration/Version31000Date20240821142813.php:
```diff
@@ -0,0 +1,36 @@
+use OCP\Migration\IOutput;
+use OCP\Migration\SimpleMigrationStep;
+
+class Version31000Date20240821142813 extends SimpleMigrationStep {
```
As this target 31 it should have a migration annotation, otherwise we have no information of the impact.

Luka-sama replied:
I have added it:
https://github.com/nextcloud/server/commit/d758442
However, my changes are no longer displayed here in this pull request, probably because the branch is already merged into master.

susnux replied:
@Luka-sama yes because it is already merged. But is also not about a doc-comment but about annotation for the kind of the migration (`OCP\Migration\Attributes\MigrationAttribute`).
Nevertheless thank you very much!!

Luka-sama replied:
@susnux Oh, I see! I must have used the older migrations as an example, the attributes seem to be something new (from the 30th version).
Should I add them too? Except that it would probably be an overkill to make a new pull request just for that 😅 (or not?)
